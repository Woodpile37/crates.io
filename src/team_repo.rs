//! The code in this module interacts with the
//! <https://github.com/rust-lang/team/> repository.
//!
//! The [TeamRepo] trait is used to abstract away the HTTP client for testing
//! purposes. The [TeamRepoImpl] struct is the actual implementation of
//! the trait.

use crate::certs;
use async_trait::async_trait;
use mockall::automock;
use reqwest::{Certificate, Client};

#[automock]
#[async_trait]
pub trait TeamRepo {
    async fn get_team(&self, name: &str) -> anyhow::Result<Team>;
}

#[derive(Debug, Clone, Deserialize)]
pub struct Team {
    pub name: String,
    pub kind: String,
    pub members: Vec<Member>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Member {
    pub name: String,
    pub github: String,
    pub github_id: i32,
    pub is_lead: bool,
}

pub struct TeamRepoImpl {
    client: Client,
}

impl TeamRepoImpl {
    fn new(client: Client) -> Self {
        TeamRepoImpl { client }
    }
}

impl Default for TeamRepoImpl {
    fn default() -> Self {
        let client = build_client();
        TeamRepoImpl::new(client)
    }
}

fn build_client() -> Client {
    let x1_cert = Certificate::from_pem(certs::ISRG_ROOT_X1).unwrap();
    let x2_cert = Certificate::from_pem(certs::ISRG_ROOT_X2).unwrap();

    Client::builder()
        .tls_built_in_root_certs(false)
        .add_root_certificate(x1_cert)
        .add_root_certificate(x2_cert)
        .build()
        .unwrap()
}

#[async_trait]
impl TeamRepo for TeamRepoImpl {
    async fn get_team(&self, name: &str) -> anyhow::Result<Team> {
        let url = format!("https://team-api.infra.rust-lang.org/v1/teams/{name}.json");
        let response = self.client.get(url).send().await?.error_for_status()?;
        Ok(response.json().await?)
    }
}

#[cfg(test)]
mod tests {
    use crate::team_repo::build_client;

    /// This test is here to make sure that the client is built
    /// correctly without panicking.
    #[test]
    fn test_build_client() {
        let _client = build_client();
    }
}
