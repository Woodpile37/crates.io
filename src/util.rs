use std::cmp;

pub use self::bytes_request::BytesRequest;
pub use self::io_util::{read_fill, read_le_u32};
pub use self::request_helpers::*;

mod bytes_request;
pub mod errors;
mod io_util;
mod request_helpers;
pub mod rfc3339;
pub mod token;
pub mod tracing;

#[derive(Debug, Copy, Clone)]
pub struct Maximums {
    pub max_upload_size: u64,
    pub max_unpack_size: u64,
}

impl Maximums {
    pub fn new(
        krate_max_upload: Option<i32>,
        app_max_upload: u64,
        app_max_unpack: u64,
    ) -> Maximums {
        let max_upload_size = krate_max_upload.map(|m| m as u64).unwrap_or(app_max_upload);
        let max_unpack_size = cmp::max(app_max_unpack, max_upload_size);
        Maximums {
            max_upload_size,
            max_unpack_size,
        }
    }
}
