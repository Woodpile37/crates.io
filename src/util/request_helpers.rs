use http::header::AsHeaderName;
use http::HeaderMap;

pub trait HeaderMapExt {
    /// Returns the value of the request header, or an empty slice if it is not
    /// present.
    ///
    /// If a header appears multiple times, this will return only one of them.
    ///
    /// If the header value is invalid utf8, an empty slice will be returned.
    fn get_str_or_default<K: AsHeaderName>(&self, key: K) -> &str;
}

impl HeaderMapExt for HeaderMap {
    fn get_str_or_default<K: AsHeaderName>(&self, key: K) -> &str {
        self.get(key)
            .map(|value| value.to_str().unwrap_or_default())
            .unwrap_or_default()
    }
}
