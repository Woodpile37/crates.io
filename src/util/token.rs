use diesel::{deserialize::FromSql, pg::Pg, serialize::ToSql, sql_types::Bytea};
use rand::{distributions::Uniform, rngs::OsRng, Rng};
use secrecy::{ExposeSecret, SecretString, SecretVec};
use sha2::{Digest, Sha256};

const TOKEN_LENGTH: usize = 32;

/// NEVER CHANGE THE PREFIX OF EXISTING TOKENS!!! Doing so will implicitly
/// revoke all the tokens, disrupting production users.
const TOKEN_PREFIX: &str = "cio";

#[derive(FromSqlRow, AsExpression)]
#[diesel(sql_type = Bytea)]
pub struct HashedToken(SecretVec<u8>);

impl HashedToken {
    pub(crate) fn parse(plaintext: &str) -> Option<Self> {
        // This will both reject tokens without a prefix and tokens of the wrong kind.
        if !plaintext.starts_with(TOKEN_PREFIX) {
            return None;
        }

        let sha256 = Self::hash(plaintext).into();
        Some(Self(sha256))
    }

    pub fn hash(plaintext: &str) -> Vec<u8> {
        Sha256::digest(plaintext.as_bytes()).as_slice().to_vec()
    }
}

impl std::fmt::Debug for HashedToken {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("HashedToken")
    }
}

impl ToSql<Bytea, Pg> for HashedToken {
    fn to_sql(&self, out: &mut diesel::serialize::Output<'_, '_, Pg>) -> diesel::serialize::Result {
        ToSql::<Bytea, Pg>::to_sql(&self.0.expose_secret(), &mut out.reborrow())
    }
}

impl FromSql<Bytea, Pg> for HashedToken {
    fn from_sql(bytes: diesel::pg::PgValue<'_>) -> diesel::deserialize::Result<Self> {
        let bytes: Vec<u8> = FromSql::<Bytea, Pg>::from_sql(bytes)?;
        Ok(Self(bytes.into()))
    }
}

#[derive(Debug)]
pub struct PlainToken(SecretString);

impl PlainToken {
    pub(crate) fn generate() -> Self {
        let plaintext = format!(
            "{}{}",
            TOKEN_PREFIX,
            generate_secure_alphanumeric_string(TOKEN_LENGTH)
        )
        .into();

        Self(plaintext)
    }

    pub fn hashed(&self) -> HashedToken {
        let sha256 = HashedToken::hash(self.expose_secret()).into();
        HashedToken(sha256)
    }
}

impl ExposeSecret<String> for PlainToken {
    fn expose_secret(&self) -> &String {
        self.0.expose_secret()
    }
}

fn generate_secure_alphanumeric_string(len: usize) -> String {
    const CHARS: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    OsRng
        .sample_iter(Uniform::from(0..CHARS.len()))
        .map(|idx| CHARS[idx] as char)
        .take(len)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use googletest::prelude::*;

    #[test]
    fn test_generated_and_parse() {
        let token = PlainToken::generate();
        assert_that!(token.expose_secret(), starts_with(TOKEN_PREFIX));
        assert_eq!(
            token.hashed().0.expose_secret(),
            Sha256::digest(token.expose_secret().as_bytes()).as_slice()
        );

        let parsed =
            HashedToken::parse(token.expose_secret()).expect("failed to parse back the token");
        assert_eq!(parsed.0.expose_secret(), token.hashed().0.expose_secret());
    }

    #[test]
    fn test_parse_no_kind() {
        assert!(HashedToken::parse("nokind").is_none());
    }
}
