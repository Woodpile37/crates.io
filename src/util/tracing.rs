use sentry::integrations::tracing::EventFilter;
use tracing::Level;
use tracing::Metadata;
use tracing_subscriber::filter::LevelFilter;
use tracing_subscriber::{prelude::*, EnvFilter};

/// Initializes the `tracing` logging framework.
///
/// Regular CLI output is influenced by the optional
/// [`RUST_LOG`](tracing_subscriber::filter::EnvFilter) environment variable
/// and is showing all `INFO` level events by default.
///
/// This function also sets up the Sentry error reporting integration for the
/// `tracing` framework, which is hardcoded to include all `INFO` level events.
pub fn init() {
    init_with_default_level(LevelFilter::INFO)
}

fn init_with_default_level(level: LevelFilter) {
    let env_filter = EnvFilter::builder()
        .with_default_directive(level.into())
        .from_env_lossy();

    let log_layer = tracing_subscriber::fmt::layer()
        .compact()
        .without_time()
        .with_filter(env_filter);

    let sentry_layer = sentry::integrations::tracing::layer()
        .event_filter(event_filter)
        .with_filter(LevelFilter::INFO);

    tracing_subscriber::registry()
        .with(log_layer)
        .with(sentry_layer)
        .init();
}

pub fn event_filter(metadata: &Metadata<'_>) -> EventFilter {
    match metadata.level() {
        &Level::ERROR if metadata.target() == "http" => EventFilter::Breadcrumb,
        &Level::ERROR => EventFilter::Exception,
        &Level::WARN | &Level::INFO => EventFilter::Breadcrumb,
        &Level::DEBUG | &Level::TRACE => EventFilter::Ignore,
    }
}

/// Initializes the `tracing` logging framework for usage in tests.
pub fn init_for_test() {
    let env_filter = EnvFilter::builder()
        .with_default_directive(LevelFilter::INFO.into())
        .from_env_lossy();

    let _ = tracing_subscriber::fmt()
        .compact()
        .with_env_filter(env_filter)
        .without_time()
        .with_test_writer()
        .try_init();
}
