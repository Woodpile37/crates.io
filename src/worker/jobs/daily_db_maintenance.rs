use crate::tasks::spawn_blocking;
use crate::worker::Environment;
use crates_io_worker::BackgroundJob;
use diesel::{sql_query, RunQueryDsl};
use std::sync::Arc;

#[derive(Serialize, Deserialize)]
pub struct DailyDbMaintenance;

impl BackgroundJob for DailyDbMaintenance {
    const JOB_NAME: &'static str = "daily_db_maintenance";

    type Context = Arc<Environment>;

    /// Run daily database maintenance tasks
    ///
    /// By default PostgreSQL will run an auto-vacuum when 20% of the tuples in a table are dead.
    /// Because the `version_downloads` table includes years of historical data, we can accumulate
    /// a *lot* of garbage before an auto-vacuum is run.
    ///
    /// We only need to keep 90 days of entries in `version_downloads`. Once we have a mechanism to
    /// archive daily download counts and drop historical data, we can drop this task and rely on
    /// auto-vacuum again.
    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
        spawn_blocking(move || {
            let mut conn = env.connection_pool.get()?;

            info!("Running VACUUM on version_downloads table");
            sql_query("VACUUM version_downloads;").execute(&mut *conn)?;
            info!("Finished running VACUUM on version_downloads table");
            Ok(())
        })
        .await
    }
}
