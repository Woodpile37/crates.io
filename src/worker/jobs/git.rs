use crate::models;
use crate::tasks::spawn_blocking;
use crate::worker::Environment;
use anyhow::Context;
use chrono::Utc;
use crates_io_env_vars::var_parsed;
use crates_io_index::{Crate, Repository};
use crates_io_worker::BackgroundJob;
use diesel::prelude::*;
use sentry::Level;
use std::fs::{self, File};
use std::io::{BufRead, BufReader, ErrorKind, Write};
use std::process::Command;
use std::sync::Arc;
use url::Url;

#[derive(Serialize, Deserialize)]
pub struct SyncToGitIndex {
    krate: String,
}

impl SyncToGitIndex {
    pub fn new(krate: impl Into<String>) -> Self {
        let krate = krate.into();
        Self { krate }
    }
}

impl BackgroundJob for SyncToGitIndex {
    const JOB_NAME: &'static str = "sync_to_git_index";
    const PRIORITY: i16 = 100;
    const QUEUE: &'static str = "repository";

    type Context = Arc<Environment>;

    /// Regenerates or removes an index file for a single crate
    #[instrument(skip_all, fields(krate.name = ? self.krate))]
    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
        info!("Syncing to git index");

        let crate_name = self.krate.clone();
        spawn_blocking(move || {
            let mut conn = env.connection_pool.get()?;
            let new = get_index_data(&crate_name, &mut conn).context("Failed to get index data")?;

            let repo = env.lock_index()?;
            let dst = repo.index_file(&crate_name);

            // Read the previous crate contents
            let old = match fs::read_to_string(&dst) {
                Ok(content) => Some(content),
                Err(error) if error.kind() == ErrorKind::NotFound => None,
                Err(error) => return Err(error.into()),
            };

            match (old, new) {
                (None, Some(new)) => {
                    fs::create_dir_all(dst.parent().unwrap())?;
                    let mut file = File::create(&dst)?;
                    file.write_all(new.as_bytes())?;
                    repo.commit_and_push(&format!("Create crate `{}`", &crate_name), &dst)?;
                }
                (Some(old), Some(new)) if old != new => {
                    let mut file = File::create(&dst)?;
                    file.write_all(new.as_bytes())?;
                    repo.commit_and_push(&format!("Update crate `{}`", &crate_name), &dst)?;
                }
                (Some(_old), None) => {
                    fs::remove_file(&dst)?;
                    repo.commit_and_push(&format!("Delete crate `{}`", &crate_name), &dst)?;
                }
                _ => debug!("Skipping sync because index is up-to-date"),
            }

            Ok(())
        })
        .await
    }
}

#[derive(Serialize, Deserialize)]
pub struct SyncToSparseIndex {
    krate: String,
}

impl SyncToSparseIndex {
    pub fn new(krate: impl Into<String>) -> Self {
        let krate = krate.into();
        Self { krate }
    }
}

impl BackgroundJob for SyncToSparseIndex {
    const JOB_NAME: &'static str = "sync_to_sparse_index";
    const PRIORITY: i16 = 100;

    type Context = Arc<Environment>;

    /// Regenerates or removes an index file for a single crate
    #[instrument(skip_all, fields(krate.name = ?self.krate))]
    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
        info!("Syncing to sparse index");

        let crate_name = self.krate.clone();
        let pool = env.connection_pool.clone();
        let content = spawn_blocking(move || {
            let mut conn = pool.get()?;
            get_index_data(&crate_name, &mut conn).context("Failed to get index data")
        })
        .await?;

        let future = env.storage.sync_index(&self.krate, content);
        future.await.context("Failed to sync index data")?;

        if let Some(cloudfront) = env.cloudfront() {
            let path = Repository::relative_index_file_for_url(&self.krate);

            info!(%path, "Invalidating index file on CloudFront");
            let future = cloudfront.invalidate(&path);
            future.await.context("Failed to invalidate CloudFront")?;
        }
        Ok(())
    }
}

#[instrument(skip_all, fields(krate.name = ?name))]
pub fn get_index_data(name: &str, conn: &mut PgConnection) -> anyhow::Result<Option<String>> {
    debug!("Looking up crate by name");
    let Some(krate): Option<models::Crate> =
        models::Crate::by_exact_name(name).first(conn).optional()?
    else {
        return Ok(None);
    };

    debug!("Gathering remaining index data");
    let crates = krate
        .index_metadata(conn)
        .context("Failed to gather index metadata")?;

    // This can sometimes happen when we delete versions upon owner request
    // but don't realize that the crate is now left with no versions at all.
    //
    // In this case we will delete the crate from the index and log a warning to
    // Sentry to clean this up in the database.
    if crates.is_empty() {
        let message = format!("Crate `{name}` has no versions left");
        sentry::capture_message(&message, Level::Warning);

        return Ok(None);
    }

    debug!("Serializing index data");
    let mut bytes = Vec::new();
    crates_io_index::write_crates(&crates, &mut bytes)
        .context("Failed to serialize index metadata")?;

    let str = String::from_utf8(bytes).context("Failed to decode index metadata as utf8")?;

    Ok(Some(str))
}

#[derive(Serialize, Deserialize)]
pub struct SquashIndex;

impl BackgroundJob for SquashIndex {
    const JOB_NAME: &'static str = "squash_index";
    const QUEUE: &'static str = "repository";

    type Context = Arc<Environment>;

    /// Collapse the index into a single commit, archiving the current history in a snapshot branch.
    #[instrument(skip_all)]
    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
        info!("Squashing the index into a single commit");

        spawn_blocking(move || {
            let repo = env.lock_index()?;

            let now = Utc::now().format("%Y-%m-%d");
            let original_head = repo.head_oid()?.to_string();
            let msg = format!("Collapse index into one commit\n\n\
            Previous HEAD was {original_head}, now on the `snapshot-{now}` branch\n\n\
            More information about this change can be found [online] and on [this issue].\n\n\
            [online]: https://internals.rust-lang.org/t/cargos-crate-index-upcoming-squash-into-one-commit/8440\n\
            [this issue]: https://github.com/rust-lang/crates-io-cargo-teams/issues/47");

            repo.squash_to_single_commit(&msg)?;

            // Shell out to git because libgit2 does not currently support push leases

            repo.run_command(Command::new("git").args([
                "push",
                // Both updates should succeed or fail together
                "--atomic",
                "origin",
                // Overwrite master, but only if it server matches the expected value
                &format!("--force-with-lease=refs/heads/master:{original_head}"),
                // The new squashed commit is pushed to master
                "HEAD:refs/heads/master",
                // The previous value of HEAD is pushed to a snapshot branch
                &format!("{original_head}:refs/heads/snapshot-{now}"),
            ]))?;

            if let Some(archive_url) = var_parsed::<Url>("GIT_ARCHIVE_REPO_URL")? {
                repo.run_command(Command::new("git").args([
                    "push",
                    archive_url.as_str(),
                    &format!("{original_head}:snapshot-{now}"),
                ]))?;
            }

            info!("The index has been successfully squashed.");

            Ok(())
        })
        .await
    }
}

#[derive(Serialize, Deserialize)]
pub struct NormalizeIndex {
    dry_run: bool,
}

impl NormalizeIndex {
    pub fn new(dry_run: bool) -> Self {
        Self { dry_run }
    }
}

impl BackgroundJob for NormalizeIndex {
    const JOB_NAME: &'static str = "normalize_index";
    const QUEUE: &'static str = "repository";

    type Context = Arc<Environment>;

    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
        info!("Normalizing the index");

        let dry_run = self.dry_run;
        spawn_blocking(move || {
            let repo = env.lock_index()?;

            let files = repo.get_files_modified_since(None)?;
            let num_files = files.len();

            for (i, file) in files.iter().enumerate() {
                if i % 50 == 0 {
                    info!(num_files, i, ?file);
                }

                let crate_name = file.file_name().unwrap().to_str().unwrap();
                let path = repo.index_file(crate_name);
                if !path.exists() {
                    continue;
                }

                let mut body: Vec<u8> = Vec::new();
                let file = fs::File::open(&path)?;
                let reader = BufReader::new(file);
                let mut versions = Vec::new();
                for line in reader.lines() {
                    let line = line?;
                    if line.is_empty() {
                        continue;
                    }

                    let mut krate: Crate = serde_json::from_str(&line)?;
                    for dep in &mut krate.deps {
                        // Remove deps with empty features
                        dep.features.retain(|d| !d.is_empty());
                        // Set null DependencyKind to Normal
                        dep.kind =
                            Some(dep.kind.unwrap_or(crates_io_index::DependencyKind::Normal));
                    }
                    krate.deps.sort();
                    versions.push(krate);
                }
                for version in versions {
                    serde_json::to_writer(&mut body, &version).unwrap();
                    body.push(b'\n');
                }
                fs::write(path, body)?;
            }

            info!("Committing normalization");
            let msg = "Normalize index format\n\n\
        More information can be found at https://github.com/rust-lang/crates.io/pull/5066";
            repo.run_command(Command::new("git").args(["commit", "-am", msg]))?;

            let branch = match dry_run {
                false => "master",
                true => "normalization-dry-run",
            };

            info!(?branch, "Pushing to upstream repository");
            repo.run_command(Command::new("git").args([
                "push",
                "origin",
                &format!("HEAD:{branch}"),
            ]))?;

            info!("Index normalization completed");

            Ok(())
        })
        .await
    }
}
