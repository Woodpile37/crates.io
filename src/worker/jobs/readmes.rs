//! Render README files to HTML.

use crate::models::Version;
use crate::tasks::spawn_blocking;
use crate::worker::Environment;
use crates_io_markdown::text_to_html;
use crates_io_worker::BackgroundJob;
use std::sync::Arc;
use tokio::runtime::Handle;

#[derive(Clone, Serialize, Deserialize)]
pub struct RenderAndUploadReadme {
    version_id: i32,
    text: String,
    readme_path: String,
    base_url: Option<String>,
    pkg_path_in_vcs: Option<String>,
}

impl RenderAndUploadReadme {
    pub fn new(
        version_id: i32,
        text: String,
        readme_path: String,
        base_url: Option<String>,
        pkg_path_in_vcs: Option<String>,
    ) -> Self {
        Self {
            version_id,
            text,
            readme_path,
            base_url,
            pkg_path_in_vcs,
        }
    }
}

impl BackgroundJob for RenderAndUploadReadme {
    const JOB_NAME: &'static str = "render_and_upload_readme";
    const PRIORITY: i16 = 50;

    type Context = Arc<Environment>;

    #[instrument(skip_all, fields(krate.name))]
    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
        use crate::schema::*;
        use diesel::prelude::*;

        info!(version_id = ?self.version_id, "Rendering README");

        let job = self.clone();
        spawn_blocking(move || {
            let rendered = text_to_html(
                &job.text,
                &job.readme_path,
                job.base_url.as_deref(),
                job.pkg_path_in_vcs.as_ref(),
            );
            if rendered.is_empty() {
                return Ok(());
            }

            let mut conn = env.connection_pool.get()?;
            conn.transaction(|conn| {
                Version::record_readme_rendering(job.version_id, conn)?;
                let (crate_name, vers): (String, String) = versions::table
                    .find(job.version_id)
                    .inner_join(crates::table)
                    .select((crates::name, versions::num))
                    .first(conn)?;

                tracing::Span::current().record("krate.name", tracing::field::display(&crate_name));

                let bytes = rendered.into();
                let future = env.storage.upload_readme(&crate_name, &vers, bytes);
                Handle::current().block_on(future)?;

                Ok(())
            })
        })
        .await
    }
}
