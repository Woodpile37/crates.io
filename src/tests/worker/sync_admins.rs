use crate::util::TestApp;
use crates_io::schema::{emails, users};
use crates_io::team_repo::{Member, MockTeamRepo, Team};
use crates_io::worker::jobs::SyncAdmins;
use crates_io_worker::BackgroundJob;
use diesel::prelude::*;
use diesel::{PgConnection, QueryResult, RunQueryDsl};
use insta::assert_debug_snapshot;
use regex::Regex;

#[test]
fn test_sync_admins_job() {
    let mock_response = mock_team(
        "crates-io",
        vec![
            mock_member("existing-admin", 1),
            mock_member("new-admin", 3),
            mock_member("new-admin-without-account", 4),
        ],
    );

    let mut team_repo = MockTeamRepo::new();
    team_repo
        .expect_get_team()
        .with(mockall::predicate::eq("crates-io-admins"))
        .returning(move |_| Ok(mock_response.clone()));

    let (app, _) = TestApp::full().with_team_repo(team_repo).empty();

    app.db(|conn| create_user("existing-admin", 1, true, conn).unwrap());
    app.db(|conn| create_user("obsolete-admin", 2, true, conn).unwrap());
    app.db(|conn| create_user("new-admin", 3, false, conn).unwrap());
    app.db(|conn| create_user("unrelated-user", 42, false, conn).unwrap());

    let admins = app.db(|conn| get_admins(conn).unwrap());
    let expected_admins = vec![("existing-admin".into(), 1), ("obsolete-admin".into(), 2)];
    assert_eq!(admins, expected_admins);

    app.db(|conn| SyncAdmins.enqueue(conn).unwrap());
    app.run_pending_background_jobs();

    let admins = app.db(|conn| get_admins(conn).unwrap());
    let expected_admins = vec![("existing-admin".into(), 1), ("new-admin".into(), 3)];
    assert_eq!(admins, expected_admins);

    let email_header_regex = Regex::new(r"(Message-ID|Date): [^\r\n]+\r\n").unwrap();
    let emails = app.as_inner().emails.mails_in_memory().unwrap();
    let emails = emails
        .iter()
        .map(|(_, email)| email_header_regex.replace_all(email, ""))
        .collect::<Vec<_>>();

    assert_debug_snapshot!(emails);

    // Run the job again to verify that no new emails are sent
    // for `new-admin-without-account`.
    app.db(|conn| SyncAdmins.enqueue(conn).unwrap());
    app.run_pending_background_jobs();

    let emails = app.as_inner().emails.mails_in_memory().unwrap();
    assert_eq!(emails.len(), 2);
}

fn mock_team(name: impl Into<String>, members: Vec<Member>) -> Team {
    Team {
        name: name.into(),
        kind: "marker-team".to_string(),
        members,
    }
}

fn mock_member(name: impl Into<String>, github_id: i32) -> Member {
    let name = name.into();
    let github = name.clone();
    Member {
        name,
        github,
        github_id,
        is_lead: false,
    }
}

fn create_user(name: &str, gh_id: i32, is_admin: bool, conn: &mut PgConnection) -> QueryResult<()> {
    let user_id = diesel::insert_into(users::table)
        .values((
            users::name.eq(name),
            users::gh_login.eq(name),
            users::gh_id.eq(gh_id),
            users::gh_access_token.eq("some random token"),
            users::is_admin.eq(is_admin),
        ))
        .returning(users::id)
        .get_result::<i32>(conn)?;

    diesel::insert_into(emails::table)
        .values((
            emails::user_id.eq(user_id),
            emails::email.eq(format!("{}@crates.io", name)),
            emails::verified.eq(true),
        ))
        .execute(conn)?;

    Ok(())
}

fn get_admins(conn: &mut PgConnection) -> QueryResult<Vec<(String, i32)>> {
    users::table
        .select((users::gh_login, users::gh_id))
        .filter(users::is_admin.eq(true))
        .order(users::gh_id.asc())
        .get_results(conn)
}
