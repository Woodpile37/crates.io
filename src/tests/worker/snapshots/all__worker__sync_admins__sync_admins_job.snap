---
source: src/tests/worker/sync_admins.rs
expression: emails
---
[
    "To: existing-admin@crates.io\r\nFrom: noreply@crates.io\r\nSubject: crates.io: Admin account changes\r\nContent-Type: text/plain; charset=utf-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nGranted admin access:\r\n\r\n- new-admin (github_id: 3)\r\n\r\nRevoked admin access:\r\n- obsolete-admin (github_id: 2)\r\n",
    "To: obsolete-admin@crates.io\r\nFrom: noreply@crates.io\r\nSubject: crates.io: Admin account changes\r\nContent-Type: text/plain; charset=utf-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nGranted admin access:\r\n\r\n- new-admin (github_id: 3)\r\n\r\nRevoked admin access:\r\n- obsolete-admin (github_id: 2)\r\n",
]
